README.txt
 GF1 (Gipf for One)
=====

  GF1 is a computer program for playing the 2 player boardgame GIPF.
  The program has a nice and easy to use graphical user interface.  You
  can play against another human (but the real game is better for that),
  against the computer, or let the computer play against itself.  The
  computer player is good enough to beat even the best human players at
  the moment (but it doesn't always win :-) ).


REMARK: Please send me a postcard if you like this program. 
        (A paper postcard, not an electronic one.)

The latest version of GF1 can be found at 
       http://gf1.sourceforge.net
The author can be reached at kurtvdb@village.uunet.be

If you have problems with this program, please let me know. I try to solve 
all problems, and all other comments and suggestions are also welcome.

contents of this file:
  + How to install
  + How to play
  + Archive contents
  + Copyright

How to install
--------------
depends on the zipfile you have:
- if the file is called
     gf1_lexe_<version>.zip or gf1_wexe_<version>.zip
  
  just unzip the file, the complete program consists of one 
  executable file (gf1 or gf1.exe).
  Just put it in any directory you like and run it.

- if the file is called
     gf1_src_<version>.zip
  then you should read the file INSTALL.txt to find out how you can
  compile the program from the source.

How to play
-----------
Just start the program and try to find out what all the menu-options do.
If that is not enough, there is a help-function.

This program does not contain an explanation of how to play GIPF. If you
don't know GIPF, go buy the game or check the GIPF-website.
(http://www.gipf.com/)

remark: the program creates a configuration file called gf1.cfg when
        you change the setup.

Archive contents
----------------
(some of these files may not be in the zipfile you have.)

COPYING.txt: information on the GNU General Public Licence
gf1_doc.txt: documentation for the program, this is the same text you can
             read when you use the help-function
HISTORY.txt: a log of all the versions I released for this program
INSTALL.txt: information on how to compile the program from the source
README.txt: what you are reading now
TODO.txt: things I still have to do to make the program even better
 
Copyright
---------
  o  GF1 (C) 1998-2000, Kurt Van den Branden.

     GF1 is free for anyone to use, but please send me a postcard if you like 
     it (a paper postcard, not an electronic one).  
     You can even send me money to support my programming efforts if you like.
     
     You can use this program for what ever you want, but don't come
     complaining to me if it explodes in your face :-).

     You can even get the source-code for this program, so you can
     change it if necessary. But if you have a useful change, please send
     it back to me so everybody can use it.

     This program is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published by
     the Free Software Foundation; either version 2 of the License, or
     (at your option) any later version.

     This program is distributed in the hope that it will be useful,
     but WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
     GNU General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with this program; if not, write to the Free Software
     Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

  o  GIPF, (C) en (R) Don & Co nv, 1997. All rights reserved.

*******************
$Id: README.txt,v 1.6 2000/02/22 18:24:39 kurt Exp $


HISTORY.txt
 v 1.03
--------
- a new way to execute a move has been added: if you now click on the
  to-position when it has already been selected, your move gets executed.
- added license-section to the about-box.
- the main window titlebar will show the name of the file you loaded or
  saved. when saving you can automatically use the same name as the 
  previously loaded file.
- started using a new filechooser widget. (should be easier for windows-users)
- the file produced by the option "save log" now has the format specified
  on the gipf website for writing down games.

 v 1.02
--------
- new saved game fileformat, everything about a game is saved now.
- possibility to save a game at any time, even at the moment when a player
  is asked what row he wants to remove or if he wants to remove a gipf
  (these changes should make it possible to use gf1 for playing by e-mail)
- the about-window now shows that the program is postcardware, so maybe
  I will receive some postcards now.
- switch to gd-1.7.3 for producing a picture of the board-situation.
  because of license-problems this library doesn't produce gif-files anymore
  but png-files instead.

 v 1.01
--------
- small performance problem: when a soft timeout occured, the best move
  from the previous iteration was used, and not from the last one.
- big performance improvement: better way of avoiding duplicate moves
  (now a lot closer in speed to version 0.99.5)
- instead of wasting time, the very first move of a game is now chosen at
  random by the computer.
- when creating a gif-file, you can now include a gipf-logo.
- I've started using STL for some lists, so when you want to compile the
  program yourself you need a compiler with a good STL implementation

 v 1.0 (not beta anymore)
-------
- solved timing-problems on MS Windows (FLTK-problem)
- new computer-player (slower than the previous one)
- computer-player can now be interrupted immediatly
- the computer-player takes time into account when playing a timed game

 v 0.99.5 beta
----------
- fixed bug: program would crash if you clicked when not on one of the
  known points (the bug was always there, but it only came to the surface
  with v 0.99.4)
- fixed bug: program would complete a computer-computer game even if you
  killed the main window
- small improvement: the logwindow will now scroll to the last line 
  automatically
- added an icon to the MS Windows version

 v 0.99.4 beta
----------
- solved problem of moves not being animated correctly after an UNDO.
- added the "Make drawing"-option to the edit-menu
  this lets you save the current board-situation as a gif-file
- internally, the resolution of the clock is now smaller than 1 second
  (in principal, it is now 1 microsecond)
    => can be important for computer-games
- GF1 should now quit correctly on MS windows when you don't use the 
  exit-menuoption. (bug reported by several people)
- checked source for compilation with fltk 1.0

 v 0.99.4.forum beta
----------------
- Only released to a very select group of people!

 v 0.99.3 beta
----------
lots of updates and improvements

- options "Start game" and "Computer move" implemented
- the documentation is reasonably complete now
- the help-window can be kept open now while playing the game.
- included patches for compilation with the Irix compiler
  (thanks to Bill Spitzak)
- The option "Save board" can be used during a game.

 v 0.99.2 beta
----------
The first version that is distributed with source-code and to the world.

The userinterface has been completely rewritten to use FLTK now instead of
V.

 v 0.91b
---------
This is a bug-fix to solve a major problem that made the computer player
very slow.

 v 0.91
--------
This is a new version to solve some bugs and add some changes
suggested by some of the users.

- changed the search-algorithm from 'mtdf' to 
  'minimax with alfa-beta pruning', this makes the searching
  slower but it looks like it plays a little better.
- give possibility to play at level 6
- several bugs that occured when the computer plays at level 1
  have been solved
- position-names are now shown on the screen
- when the computer finds several moves with the same value, he
  will choose one at random. 
- the default when starting a new game is now "tournament"

Several people have asked for an undo-button.
I'm still thinking about it, but I don't think I will implement it.
It feels too much like cheating to me.
If you've done a bad move, you've done a bad move and you have to
play on from there. If you're playing against another player, you
can not take back a move either.
(If somebody has another point of view, please let me know)

This version is still not ver general distribution.
Only people that get the program directly from me should have it.

 v 0.90
--------
First version to be distributed to a select group of people.
The program is not ready for general distribution yet.

*************************************************
$Id: HISTORY.txt,v 1.8 2000/02/28 19:48:31 kurt Exp $


xmlite.h
/*
** $Id: xmlite.h,v 1.2 2000/01/25 19:02:28 kurt Exp $
**
** code for reading and writing data in an xml-like format
**
** the output from these routines may look a lot like xml, and may even
** be wel-formed xml, but I don't think you should use it as that.
*/
/*
**    Copyright (C) 2000 Kurt Van den Branden
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
** 
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
** 
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.  
*/

/*
** the following is an example of what you should be able to store:

<gipfgame type="tournament">

<whiteplayer type="human" fulltime="1200" timeleft="1197.562988" gipf="yes">
White</whiteplayer>
<blackplayer type="computer" fulltime="1200" timeleft="1196" gipf="no">
Black</blackplayer>
<nextplayer>white</nextplayer>

<!-- in between state of removing rows or gipf-pieces -->

<board>
<whitepieces>16</whitepieces>
<whitelost>0</whitelost>
<blackpieces>16</blackpieces>
<blacklost>0</blacklost>
<position color="white" type="piece">e2</position>
<position color="black" type="gipf">d5</position>
</board>

<gamelog>
<move nr="1" player="white" type="gipf">e1-e2</move>
<move nr="1" player="black" type="piece">e1-e3</move>
<removerow owner="white">
<position color="black" type="piece">e2</position>
<position color="white" type="piece">e3</position>
<position color="white" type="piece">e4</position>
<position color="white" type="gipf">e5</position>
</removerow>
</gamelog>

</gipfgame>

*/

#ifndef _XMLITE_H_
#define _XMLITE_H_ 1

#include <string>
#include <hash_map>
#include <vector>

struct eqstring
{
    bool operator()(const string & s1, const string & s2) const
	{
	    return (s1 == s2);
	}
};


struct stringhash
{
    bool operator()(const string & s) const
	{
	    hash<const char *> H;
	    return (H (s.c_str()));
	}
};


class xmlite_entity {
 public:
    xmlite_entity (const char * ename);
    xmlite_entity (const string & ename);
    xmlite_entity (const xmlite_entity & source)
	{ (*this) = source; };
    ~xmlite_entity ();

    xmlite_entity & operator= (const xmlite_entity & source);

    int addcontent (xmlite_entity * newcontent);
    int addattribute (const string & aname, const string & avalue);
    int setvalue (const string & newvalue);

    int writetofile (const char * filename);

    string * escapechars (const string & instring);
    void escapechars (string & instring);
    string * unescapechars (const string & instring);
    void unescapechars (string & instring);

    xmlite_entity * getcontentbynr (const int nr = 0);
    xmlite_entity * getcontentbyname (const string & name,
				      const int nr = 0);
    const string & getvalue ()
	{ return (value); };
    const string & getname ()
	{ return (name); };
    const string & getattribute (const string key)
	{ return (attributes[key]); };
    
 private:
    string name;
    string value;
    vector<xmlite_entity *> contents;
    hash_map<string, string, stringhash, eqstring> attributes;

    int writetostream (ostream & ostr);
    void replaceall (string & input, const string & from, const string & to);
};


class xmlite_parser {
 public:
    xmlite_parser (istream * inputstream) :
	mystream (false), istr (inputstream)
	{ parseerror = "no parsing done yet."; };
    xmlite_parser (const char * filename);
    ~xmlite_parser ()
	{ if (mystream) delete (istr); };

    xmlite_entity * parse ();
    xmlite_entity * readentity (const string & startname);
    void getnextc ();
    void skipwhitespace ();
    void readstring (string & str, const char * endchar);
    void readattribute (xmlite_entity * ent);
    
    string & geterror ()
	{ return (parseerror); };
    
 private:
    bool mystream;
    istream * istr;
    char nextc;
    
    string parseerror;
};

#endif
